src/utils.zig

```diff
@@ -41,7 +41,18 @@ pub fn List(comptime T: type, comptime N: usize) type {
 
         pub fn sszDecode(serialized: []const u8, out: *Self, allocator: ?std.mem.Allocator) !void {
             // BitList[N] or regular List[N]?
-            out.* = try init(0);
+            const alloc = allocator orelse return error.AllocatorRequired;
+            out.* = try init(alloc);
+
+            // FastSSZ-style capacity optimization: pre-allocate based on input size
+            if (serialized.len > 0) {
+                const estimated_capacity = if (try lib.isFixedSizeObject(Self.Item))
+                    serialized.len / (try lib.serializedFixedSize(Self.Item))
+                else
+                    serialized.len / 8; // Conservative estimate for dynamic types
+                try out.inner.ensureTotalCapacity(estimated_capacity);
+            }
+
```

Comment:
1. fast-ssz is buggy, don't take it as a reference
2. this is so you can allocate 4000 bytes items, the `/8` heuristic is most likely wrong
3.  later on, you call `decodeDynamicListLength`, you should ensure the capacity at this point instead of guessing and hoping that this will not cause any further allocations
4. conversely, you could allocate too much memory for nothing.